cli/printlog/compact/compact.go

```diff
@@ -0,0 +1,267 @@
+// This format is still experimental and could be changed in the future.
+package compact
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"path"
+	"slices"
+	"strings"
+
+	"github.com/buildbuddy-io/buildbuddy/cli/log"
+	"google.golang.org/protobuf/encoding/protodelim"
+
+	spb "github.com/buildbuddy-io/buildbuddy/proto/spawn"
+)
+
+// SpawnLogReconstructor reconstructs "compact execution log" format back to the original format.
+// As of Bazel 7.1, this is the recommended way to consume the new compact format.
+type SpawnLogReconstructor struct {
+	input *bufio.Reader
+
+	hashFunc string
+	files    map[int32]*spb.File
+	dirs     map[int32]*reconstructedDir
+	symlinks map[int32]*spb.File
+	sets     map[int32]*spb.ExecLogEntry_InputSet
+
+	sort bool
+}
+
+type reconstructedDir struct {
+	path  string
+	files []*spb.File
+}
+
+func NewSpawnLogReconstructor(input io.Reader, sort bool) *SpawnLogReconstructor {
+	return &SpawnLogReconstructor{
+		input:    bufio.NewReader(input),
+		hashFunc: "",
+		files:    make(map[int32]*spb.File),
+		dirs:     make(map[int32]*reconstructedDir),
+		symlinks: make(map[int32]*spb.File),
+		sets:     make(map[int32]*spb.ExecLogEntry_InputSet),
+		sort:     sort,
+	}
+}
+
+func (slr *SpawnLogReconstructor) GetSpawnExec() (*spb.SpawnExec, error) {
+	entry := &spb.ExecLogEntry{}
+	for {
+		err := protodelim.UnmarshalFrom(slr.input, entry)
+		if err == io.EOF {
+			return nil, io.EOF
+		}
+		if err != nil {
+			return nil, fmt.Errorf("failed to read execution log entry: %s", err)
+		}
+
+		switch e := entry.GetType().(type) {
+		case *spb.ExecLogEntry_Invocation_:
+			slr.hashFunc = e.Invocation.GetHashFunctionName()
+		case *spb.ExecLogEntry_File_:
+			slr.files[entry.GetId()] = reconstructFile(nil, e.File)
+		case *spb.ExecLogEntry_Directory_:
+			slr.dirs[entry.GetId()] = reconstructDir(e.Directory)
+		case *spb.ExecLogEntry_UnresolvedSymlink_:
+			slr.symlinks[entry.GetId()] = reconstructSymlink(e.UnresolvedSymlink)
+		case *spb.ExecLogEntry_InputSet_:
+			slr.sets[entry.GetId()] = e.InputSet
+		case *spb.ExecLogEntry_Spawn_:
+			return slr.reconstructSpawn(e.Spawn)
+		default:
+			log.Warnf("unknown exec log entry: %v", entry)
+		}
+	}
+}
+
+func (slr *SpawnLogReconstructor) reconstructSpawn(s *spb.ExecLogEntry_Spawn) (*spb.SpawnExec, error) {
+	se := &spb.SpawnExec{
+		CommandArgs:     s.GetArgs(),
+		TargetLabel:     s.GetTargetLabel(),
+		Mnemonic:        s.GetMnemonic(),
+		ExitCode:        s.GetExitCode(),
+		Status:          s.GetStatus(),
+		Runner:          s.GetRunner(),
+		CacheHit:        s.GetCacheHit(),
+		Remotable:       s.GetRemotable(),
+		Cacheable:       s.GetCacheable(),
+		RemoteCacheable: s.GetRemoteCacheable(),
+		TimeoutMillis:   s.GetTimeoutMillis(),
+		Metrics:         s.GetMetrics(),
+		Digest:          s.GetDigest(),
+	}
+	// Handle env vars
+	envVar := s.GetEnvVars()
+	if slr.sort {
+		slices.SortFunc(envVar, compMapEntries)
```

Comment:
The environment variables are already sorted: https://github.com/bazelbuild/bazel/blob/62f302b4f1205bf1769bb7fbb5eff1e62f51dca2/src/main/java/com/google/devtools/build/lib/exec/SpawnLogContext.java#L85

Reply:
will remove this